### PR TITLE
[memprof] Add MemProf version

### PR DESCRIPTION
**Pull request by kazutakahirata**
This patch adds a version field to the MemProf section of the indexed
profile format, calling the new version "version 1".  The existing
version is called "version 0".

The writer supports both versions via a command-line option:

  llvm-profdata merge --memprof-version=1 ...

The reader supports both versions by automatically detecting the
version from the header.

### CODE REVIEW

File: llvm/include/llvm/ProfileData/MemProf.h [hunk 1]
```diff
@@ -15,6 +15,15 @@
 namespace llvm {
 namespace memprof {
 
+// The versions of the indexed MemProf format
+enum MemProfVersion {
```
Comment (snehasish) [inline llvm/include/llvm/ProfileData/MemProf.h:19]: Perhaps call this IndexedVersion so when we spell it out with the namespace (e.g. `memprof::IndexedVersion`) it's clear to the reader which format we are referring to?

File: llvm/include/llvm/ProfileData/InstrProfWriter.h [hunk 2]
```diff
@@ -75,11 +75,16 @@ class InstrProfWriter {
...
 
+  // The version we should use to write MemProf in.
+  memprof::MemProfVersion MemProfVersionRequested;
+
 public:
   InstrProfWriter(bool Sparse = false,
                   uint64_t TemporalProfTraceReservoirSize = 0,
                   uint64_t MaxTemporalProfTraceLength = 0,
-                  bool WritePrevVersion = false);
+                  bool WritePrevVersion = false,
```
Comment (snehasish) [inline llvm/include/llvm/ProfileData/InstrProfWriter.h:85]: Do we need WritePrevVersion? I don't see it being used in the patch. External APIs can directly specify the version once this patch lands so I think WritePrevVersion may be dropped?

File: llvm/lib/ProfileData/InstrProfReader.cpp [hunk 3]
```diff
@@ -1230,10 +1231,38 @@ Error IndexedInstrProfReader::readHeader() {
...
+    } else if (FirstWord >= 24) {
+      // This is a heuristic/hack to detect memprof::MemProfVersion0,
+      // which does not have a version field in the header.
+      // In memprof::MemProfVersion0, FirstWord should be RecordTableOffset,
+      // which should be at least 24 because of the MemProf header size.
+      Version = memprof::MemProfVersion0;
+    } else {
+      return make_error<InstrProfError>(
+          instrprof_error::unsupported_version,
+          formatv("MemProf version {} not supported; version 0 or 1 required",
```
Comment (snehasish) [inline llvm/lib/ProfileData/InstrProfReader.cpp:1256]: Can we frame this error in a way that we don't need to change the string everytime we update the version?

File: llvm/lib/ProfileData/InstrProfWriter.cpp [hunk 4]
```diff
@@ -528,7 +529,15 @@ Error InstrProfWriter::writeImpl(ProfOStream &OS) {
   // OnDiskChainedHashTable MemProfFrameData
   uint64_t MemProfSectionStart = 0;
   if (static_cast<bool>(ProfileKind & InstrProfKind::MemProf)) {
+    assert((MemProfVersionRequested == memprof::MemProfVersion0 ||
```
Comment (snehasish) [inline llvm/lib/ProfileData/InstrProfWriter.cpp:532]: I think this should raise an actual error instead of just an assert.

File: llvm/include/llvm/ProfileData/InstrProfWriter.h (hunk 2 above)
Comment (teresajohnson) [reply llvm/include/llvm/ProfileData/InstrProfWriter.h:85]: This is not a change in this patch, and we should remove it separately.

File: llvm/include/llvm/ProfileData/InstrProfWriter.h [hunk 5]
```diff
@@ -75,11 +75,16 @@ class InstrProfWriter {
   // deployment of newer versions of llvm-profdata.
   bool WritePrevVersion = false;
 
+  // The version we should use to write MemProf in.
```
Comment (teresajohnson) [inline llvm/include/llvm/ProfileData/InstrProfWriter.h:78]: Maybe "The MemProf version we should write"? (the "write MemProf in" at the end is a little hard to parse)

File: llvm/include/llvm/ProfileData/MemProf.h [hunk 6]
```diff
@@ -15,6 +15,15 @@
...
+  // Version 0: This version didn't have a version field have a version in the
```
Comment (teresajohnson) [inline llvm/include/llvm/ProfileData/MemProf.h:20]: Some extra words here ("have a version field have a version")

File: llvm/lib/ProfileData/InstrProfReader.cpp [hunk 7]
```diff
@@ -1230,10 +1231,38 @@ Error IndexedInstrProfReader::readHeader() {
...
+
+    memprof::MemProfVersion Version = memprof::MemProfVersion0;
+    if (static_cast<memprof::MemProfVersion>(FirstWord) ==
+        memprof::MemProfVersion1) {
+      // Everything is good.  We can proceed to deserialize the rest.
+      Version = memprof::MemProfVersion1;
+    } else if (FirstWord >= 24) {
+      // This is a heuristic/hack to detect memprof::MemProfVersion0,
+      // which does not have a version field in the header.
+      // In memprof::MemProfVersion0, FirstWord should be RecordTableOffset,
```
Comment (teresajohnson) [inline llvm/lib/ProfileData/InstrProfReader.cpp:1250]: "will be RecordTableOffset"

File: llvm/lib/ProfileData/InstrProfWriter.cpp [hunk 8]
```diff
@@ -528,7 +529,15 @@ Error InstrProfWriter::writeImpl(ProfOStream &OS) {
...
+            MemProfVersionRequested == memprof::MemProfVersion1) &&
+           "Requested MemProf version not supported");
+
     MemProfSectionStart = OS.tell();
+
+    if (MemProfVersionRequested == memprof::MemProfVersion1)
```
Comment (teresajohnson) [inline llvm/lib/ProfileData/InstrProfWriter.cpp:538]: Maybe generalize this so it doesn't have to be tweaked in future versions:

```
if (MemProfVersionRequested >= memprof::MemProfVersion1)
   OS.write(MemProfVersionRequested);
```

File: llvm/lib/ProfileData/InstrProfWriter.cpp [hunk 9]
```diff
@@ -570,12 +579,13 @@ Error InstrProfWriter::writeImpl(ProfOStream &OS) {
...
-        {MemProfSectionStart, &RecordTableOffset, 1},
-        {MemProfSectionStart + sizeof(uint64_t), &FramePayloadOffset, 1},
-        {MemProfSectionStart + 2 * sizeof(uint64_t), &FrameTableOffset, 1},
-    };
-    OS.patch(PatchItems);
+    uint64_t Header[] = {RecordTableOffset, FramePayloadOffset,
+                         FrameTableOffset};
+    uint64_t HeaderUpdatePos = MemProfSectionStart;
+    if (MemProfVersionRequested == memprof::MemProfVersion1)
+      // The updates goes just after the version field.
```
Comment (teresajohnson) [inline llvm/lib/ProfileData/InstrProfWriter.cpp:586]: "updates go"

File: llvm/lib/ProfileData/InstrProfWriter.cpp [hunk 10]
```diff
@@ -570,12 +579,13 @@ Error InstrProfWriter::writeImpl(ProfOStream &OS) {
...
-    PatchItem PatchItems[] = {
-        {MemProfSectionStart, &RecordTableOffset, 1},
-        {MemProfSectionStart + sizeof(uint64_t), &FramePayloadOffset, 1},
-        {MemProfSectionStart + 2 * sizeof(uint64_t), &FrameTableOffset, 1},
-    };
-    OS.patch(PatchItems);
+    uint64_t Header[] = {RecordTableOffset, FramePayloadOffset,
+                         FrameTableOffset};
+    uint64_t HeaderUpdatePos = MemProfSectionStart;
+    if (MemProfVersionRequested == memprof::MemProfVersion1)
```
Comment (teresajohnson) [inline llvm/lib/ProfileData/InstrProfWriter.cpp:585]: Similar to above, make this `if (MemProfVersionRequested >= memprof::MemProfVersion1)` so it doesn't need to be changed in the future.

**Comment by kazutakahirata** [comment]
Thank you for the thorough reviews!  Please take a look at the revised version.

File: llvm/include/llvm/ProfileData/InstrProfWriter.h [hunk 11]
```diff
@@ -75,16 +75,14 @@ class InstrProfWriter {
   // deployment of newer versions of llvm-profdata.
   bool WritePrevVersion = false;
 
-  // The version we should use to write MemProf in.
-  memprof::MemProfVersion MemProfVersionRequested;
+  // The MemProf version we should use to write in.
```
Comment (teresajohnson) [inline llvm/include/llvm/ProfileData/InstrProfWriter.h:78]: Suggest replacing "use to write in" with simply "write"

**Review by teresajohnson — APPROVED** [review]
One minor comment nit below. Wait to see if @snehasish has more comments though, but lgtm.

**Review by snehasish — APPROVED** [review]
LGTM. This change seems simple enough but we should have some way of testing reading / writing new and old versions before we make larger changes.

**Comment by dyung** [comment]
Hi @kazutakahirata, our internal Windows builder (Visual Studio 2019, Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29924 for x64) hits a compilation error when trying to build your change:
```
C:\j\w\779ddbee\p\llvm\include\llvm/Support/FormatProviders.h(129): error C2872: 'detail': ambiguous symbol
C:\j\w\779ddbee\p\llvm\include\llvm/ADT/Sequence.h(110): note: could be 'llvm::detail'
C:\j\w\779ddbee\p\llvm\include\llvm/Support/Endian.h(197): note: or       'llvm::support::detail'
C:\j\w\779ddbee\p\llvm\include\llvm/Support/FormatVariadicDetails.h(73): note: see reference to class template instantiation 'llvm::format_provider<unsigned __int64,void>' being compiled
C:\j\w\779ddbee\p\llvm\include\llvm/Support/FormatVariadicDetails.h(106): note: see reference to class template instantiation 'llvm::detail::has_FormatProvider<T>' being compiled
        with
        [
            T=const uint64_t &
        ]
C:\j\w\779ddbee\p\llvm\include\llvm/Support/FormatVariadicDetails.h(124): note: see reference to class template instantiation 'llvm::detail::uses_format_provider<T>' being compiled
        with
        [
            T=const uint64_t &
        ]
C:\j\w\779ddbee\p\llvm\include\llvm/Support/FormatVariadic.h(252): note: see reference to class template instantiation 'llvm::detail::uses_missing_provider<const uint64_t &>' being compiled
C:\j\w\779ddbee\p\llvm\include\llvm/Support/FormatVariadicDetails.h(156): note: see reference to function template instantiation 'enable_if<llvm::detail::uses_missing_provider<T>::value,llvm::detail::missing_format_adapter<T>>::type llvm::detail::build_format_adapter(T &&)' being compiled
C:\j\w\779ddbee\p\llvm\lib\ProfileData\InstrProfReader.cpp(1255): error C2672: 'llvm::formatv': no matching overloaded function found
C:\j\w\779ddbee\p\llvm\lib\ProfileData\InstrProfReader.cpp(1258): error C2893: Failed to specialize function template 'llvm::formatv_object<unknown-type> llvm::formatv(const char *,Ts &&...)'
C:\j\w\779ddbee\p\llvm\include\llvm/Support/FormatVariadic.h(251): note: see declaration of 'llvm::formatv'
C:\j\w\779ddbee\p\llvm\lib\ProfileData\InstrProfReader.cpp(1258): note: With the following template arguments:
C:\j\w\779ddbee\p\llvm\lib\ProfileData\InstrProfReader.cpp(1258): note: 'Ts={const uint64_t &, const uint64_t &, const uint64_t &}'
```
Can you take a look?

**Comment by kazutakahirata** [comment]
> Can you take a look?

Ack.  Thank you for reporting this!

**Comment by dyung** [comment]
> > Can you take a look?
> 
> @dyung Do you get similar errors on `llvm/lib/ProfileData/InstrProfWriter.cpp`?
> 
> At `llvm/lib/ProfileData/InstrProfReader.cpp:1255`, the types of the arguments to `formatv` are:
> 
> 1. `const char[80]`
> 2. `const uint64_t`
> 3. `const uint64_t`
> 4. `const uint64_t`
> 
> At `llvm/lib/ProfileData/InstrProfWriter.cpp:538`, the types of the arguments to `formatv` are:
> 
> 1. `const char[80]`
> 2. `memprof::IndexedVersion` (whose underlying type is `uint64_t`)
> 3. `const uint64_t`
> 4. `const uint64_t`
> 
> I am guessing that if `const uint64_t` is problematic, `InstrProfWriter.cpp` should trigger similar errors.
> 
> With all that said, it's possible for your build system to terminate the build upon the first error. It may not have a chance to get to `InstrProfWriter.cpp` if it encounters an error with `InstrProfReader.cpp` first.

We do see the same errors with `llvm/lib/ProfileData/InstrProfWriter.cpp`, I left it out since fixing one should fix both I think, sorry about that.

I was able to work around the issue with `llvm::detail` by explicitly specifying `llvm::detail` at line `llvm/include/llvm/Support/FormatProviders.h:129`.
```diff
diff --git a/llvm/include/llvm/Support/FormatProviders.h b/llvm/include/llvm/Support/FormatProviders.h
index aa0773847161..a83179f014b1 100644
--- a/llvm/include/llvm/Support/FormatProviders.h
+++ b/llvm/include/llvm/Support/FormatProviders.h
@@ -126,7 +126,7 @@ protected:
 template <typename T>
 struct format_provider<
     T, std::enable_if_t<detail::use_integral_formatter<T>::value>>
-    : public detail::HelperFunctions {
+    : public llvm::detail::HelperFunctions {
 private:
 public:
   static void format(const T &V, llvm::raw_ostream &Stream, StringRef Style) {
```
I'm still trying to figure out how to get the compiler to find the correct overload for `formatv` though.

**Comment by kazutakahirata** [comment]
> We do see the same errors with `llvm/lib/ProfileData/InstrProfWriter.cpp`, I left it out since fixing one should fix both I think, sorry about that.

No apologies needed.  I was just wondering about the differences between the two call sites of `formatv`.

> ```diff
> diff --git a/llvm/include/llvm/Support/FormatProviders.h b/llvm/include/llvm/Support/FormatProviders.h
> index aa0773847161..a83179f014b1 100644
> --- a/llvm/include/llvm/Support/FormatProviders.h
> +++ b/llvm/include/llvm/Support/FormatProviders.h
> @@ -126,7 +126,7 @@ protected:
>  template <typename T>
>  struct format_provider<
>      T, std::enable_if_t<detail::use_integral_formatter<T>::value>>
> -    : public detail::HelperFunctions {
> +    : public llvm::detail::HelperFunctions {

@dyung Wow, thank you so much for looking into this.  Adding `llvm::` may be a good practical fix if the host compiler doesn't find the correct overload.  Would you like to create a PR for that?

> I'm still trying to figure out how to get the compiler to find the correct overload for `formatv` though.

So, the `llvm::` above isn't sufficient to fix the build?

**Comment by dyung** [comment]
> > We do see the same errors with `llvm/lib/ProfileData/InstrProfWriter.cpp`, I left it out since fixing one should fix both I think, sorry about that.
> 
> No apologies needed. I was just wondering about the differences between the two call sites of `formatv`.
> 
> > ```diff
> > diff --git a/llvm/include/llvm/Support/FormatProviders.h b/llvm/include/llvm/Support/FormatProviders.h
> > index aa0773847161..a83179f014b1 100644
> > --- a/llvm/include/llvm/Support/FormatProviders.h
> > +++ b/llvm/include/llvm/Support/FormatProviders.h
> > @@ -126,7 +126,7 @@ protected:
> >  template <typename T>
> >  struct format_provider<
> >      T, std::enable_if_t<detail::use_integral_formatter<T>::value>>
> > -    : public detail::HelperFunctions {
> > +    : public llvm::detail::HelperFunctions {
> > ```
> 
> @dyung Wow, thank you so much for looking into this. Adding `llvm::` may be a good practical fix if the host compiler doesn't find the correct overload. Would you like to create a PR for that?
> 
> > I'm still trying to figure out how to get the compiler to find the correct overload for `formatv` though.
> 
> So, the `llvm::` above isn't sufficient to fix the build?

It does not unfortunately. It seems there were two different, somewhat related issues. The first was the ambiguous llvm::detail, and the inability to resolve the call to `formatv`. The change I proposed fixes the former, but the latter is still there.

**Comment by dyung** [comment]
I believe we found the issue with our build configuration. A while back we had disabled the option `/permissive-` for our internal builds due to a bug in the version of Visual Studio that we are using internally. The version has since been updated, so I tried re-enabling that switch and I can now build your change on our internal Windows builder. So it looks like it was due to some differences in how the compiler interpreted the standard here. So with that our internal builder is working again and sorry for the confusion!

**Comment by kazutakahirata** [comment]
> I can now build your change on our internal Windows builder.

I'm glad you were able to restore the build.
